cdx_vietnam/middleware/cdp_client.rb

```
@@ -0,0 +1,110 @@
+
+# This class is base class for connect api
+class CdpClient
+  attr_accessor :max_retry
+  def initialize
+    @mec = Mechanize.new
+    @token = ""
+    @sleeping_times = [2,2,1]
+    max_retry = 5
+    # base_url: root host address. Example: google.com
+    @base_url = ""
+    @login_url = ""
+    @get_patient_url = ""
+    @create_patient_url = ""
+    @update_patient_url = ""
+    @create_exam_request_url = ""
+    @update_exam_request_url = ""
+    @search_patient_url = ""
+    @is_use_token = false
+  end
+
+  # Login method for authentication
+  # *Params* :
+  #  - url: full login path
+  #  - user_name: username for login
+  #  - password: password to login
+  # *Return* : none
+  def login(url, user_name, password)
+    @url = base_url
+    @user_name = user_name
+    @pass = password
+    authenticate
+  end
+
+  # Post method to post data to api server and handle error when exception (timeout, api server down, ...) occur
+  # and retry several times
+  # *Params* :
+  #  - url: full url to post data
+  #  - data: data to post
+  # *Return* : api-server's respond
+  def post_with_handle(url, data)
+    data = data_with_token(data) if @is_use_token
+    count_retry = 0
+    begin
+      resp = @mec.post(url, data)
+      return resp
+    rescue Exception => e
+      if count_retry < @sleeping_times.count
+        sleep(@sleeping_times[count_retry])
+      end
+      log(e.message + url + data.to_s)
+      count_retry += 1
+      retry if count_retry < max_retry
+    end
+  end
+
+  # Get method to get data from api server and handle error when exception (timeout, api server down, ...) occur
+  # and retry several times
+  # *Params* :
+  #  - url: full url to get data
+  #  - data: data will be send to api server
+  # *Return* : api-server's respond
+  def get_with_handle(url, data)
+    data = data_with_token(data) if @is_use_token
+    count_retry = 0
+    begin
+      resp = @mec.get(url, data)
+      return resp
+    rescue Exception => e
```

Never rescue `Exception` (ever) if you must then you should re-raise the error. If you must rescue random exceptions use `StandardError` instead. Randomly rescuing `Exception` will rescue ALL errors that inherit from `Exception` and could lead to some behaviour that we never ever want to see such as `NoMemoryError` and I don't think we want to see gracefully rescue an error in production that is going to eat all the RAM do we???